File: 2300-successful-pairs-of-spells-and-potions/2300-successful-pairs-of-spells-and-potions.java
class Solution {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        int n = potions.length;
        int m = spells.length;
        int[] result = new int[m];
        Arrays.sort(potions);

        for (int i = 0; i < m ; i++) {
            int lo = 0, hi = n - 1;

            while (lo <= hi) {
                int mid = lo + (hi - lo) / 2;
                long chk = (long)potions[mid] * spells[i] - success;
                if (chk >= 0) {
                    hi = mid - 1;
                } else {
                    lo = mid + 1;
                }
            }

            result[i] = n - lo;
        }

        return result;
    }
}

File: 2300-successful-pairs-of-spells-and-potions/2300-successful-pairs-of-spells-and-potions.py
class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        result = list()
        potions.sort()
        n = len(potions)

        for spell in spells:
            lo, hi = 0, n - 1

            while lo <= hi:
                mid = (lo + hi) // 2
                chk = potions[mid] * spell - success
                if chk >= 0:
                    hi = mid - 1
                else:
                    lo = mid + 1

            result.append(n - lo)

        return result